noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr

```diff
@@ -0,0 +1,148 @@
+use dep::protocol_types::{
+    address::AztecAddress, constants::GENERATOR_INDEX__NOTE_HASH,
+    hash::poseidon2_hash_with_separator, traits::Hash,
+};
+
+use crate::{
+    context::{PrivateContext, UtilityContext},
+    keys::getters::{get_nsk_app, get_public_keys},
+    oracle::notes::check_nullifier_exists,
+    state_vars::owned_state_variable::OwnedStateVariable,
+};
+
+mod test;
+
+/// A private state variable type that represents a right a user can exercise at most once.
+///
+/// ## Example
+///
+/// The right to vote at most once can be implemented using [SingleUseClaim] as a primitive.
+///
+/// ```noir
+/// #[derive(Serialize, Deserialize)]
+/// struct ElectionId { id: Field }
+///
+/// #[storage]
+/// struct Storage<Context> {
+///     tally: Map<ElectionId, Map<Field, PublicMutable<Field, Context>, Context>, Context>, // election => candidate => number of votes
+///     vote_claims: Map<ElectionId, Owned<SingleUseClaim<Context>, Context>, Context>, // track right to vote at most once per election
+/// }
+///
+/// #[external("private")]
+/// fn cast_vote(election_id: ElectionId, candidate: Field) {
+///     let voter = self.msg_sender().unwrap();
+///
+///     // `claim` will only succeed if `voter` hasn't ever run this before.
+///     self.storage.vote_claims.at(election_id).at(voter).claim();
+///
+///     // Now that we checked that the voter is entitled to vote,
+///     // we can enqueue the corresponding public effect
+///     self.enqueue_self.add_to_tally_public(election_id, candidate);
+/// }
+/// ```
+///
+/// [SingleUseClaim] state variables are declared within a contract's `#[storage]` struct:
+///
+/// `your_claimable: SingleUseClaim<Context>`
+///
+/// ## Cost
+///
+/// Exercising a claim (via the [claim] function) emits one nullifier.
+///
+/// ## Privacy considerations
+///
+/// The [claim] function emits a public nullifier per `owner`:
+/// this leaks the fact that _someone_ exercised that claim once, but the owner is hidden
```

Comment:
> this leaks the fact that _someone_ exercised that claim once

Technically, the act of emitting the nullifier doesn't leak that a claim was executed (or even that this contract was executed). It's the other app-specific logic (e.g. the public effects that you allude to a few lines later) that might leak the fact that someone exercised a claim.